src/parser/shaman/elemental/modules/talents/Ascendance.tsx

```diff
@@ -0,0 +1,144 @@
+import React from 'react';
+
+import SPELLS from 'common/SPELLS';
+import { formatNumber, formatPercentage } from 'common/format';
+
+import Analyzer, { Options } from 'parser/core/Analyzer';
+import EnemyInstances from 'parser/shared/modules/EnemyInstances';
+
+import { STATISTIC_ORDER } from 'interface/others/StatisticBox';
+
+import Abilities from 'parser/shaman/elemental/modules/Abilities';
+import Statistic from 'interface/statistics/Statistic';
+import STATISTIC_CATEGORY from 'interface/others/STATISTIC_CATEGORY';
+import { CastEvent } from 'parser/core/Events';
+import Enemies from 'parser/shared/modules/Enemies';
+import BoringSpellValueText from 'interface/statistics/components/BoringSpellValueText';
+import { ThresholdStyle, When } from 'parser/core/ParseResults';
+
+const ASCENDANCE_DURATION = 15000 - 1500; //remove the gcd for Ascendence itself because we only check for FS on the first cast after
+
+class Ascendance extends Analyzer {
+
+  static dependencies = {
+    abilities: Abilities,
+    enemies: EnemyInstances,
+  };
+
+  protected abilities!: Abilities;
+  protected enemies!: Enemies;
+
+  justEnteredAscendance: boolean = false;
+  badFSAscendence: number = 0;
+  checkDelay: number = 0;
+
+  numCasts = {
+    [SPELLS.ASCENDANCE_TALENT_ELEMENTAL.id]: 0,
+    [SPELLS.LAVA_BURST.id]: 0,
+    [SPELLS.EARTH_SHOCK.id]: 0,
+    [SPELLS.ELEMENTAL_BLAST_TALENT.id]: 0,
+    others: 0,
+  };
+
+  constructor(options: Options) {
+    super(options);
+    this.active = this.selectedCombatant.hasTalent(SPELLS.ASCENDANCE_TALENT_ELEMENTAL.id);
+  }
+
+  get AscendanceUptime() {
+    return this.selectedCombatant.getBuffUptime(SPELLS.ASCENDANCE_TALENT_ELEMENTAL.id) / this.owner.fightDuration;
+  }
+
+  get averageLavaBurstCasts() {
+    return (this.numCasts[SPELLS.LAVA_BURST.id] / this.numCasts[SPELLS.ASCENDANCE_TALENT_ELEMENTAL.id]) || 0;
+  }
+
+  on_byPlayer_cast(event: CastEvent) {
+    const spellId = event.ability.guid;
+    const target = this.enemies.getEntity(event);
+
+    if (spellId === SPELLS.ASCENDANCE_TALENT_ELEMENTAL.id) {
+      this.justEnteredAscendance = true;
+      this.numCasts[SPELLS.ASCENDANCE_TALENT_ELEMENTAL.id] += 1;
+    }
+
+    const ability = this.abilities.getAbility(spellId);
+    if (!ability) {
+      return;
+    }
+
```

Comment:
What are you using ability for here? It's never used for further below in this function